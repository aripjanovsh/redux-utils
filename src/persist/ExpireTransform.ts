import { invariant, isFunction, isNumber } from "../internal/utils";
import { TransformConfig } from "./PersistConfigBuilder";

interface Plain<S> {
  r: S;
  t: number;
}

export function createExpireTransformConfig<S>(
  expireAfter: number,
  now: () => number = Date.now,
): TransformConfig<S, Plain<S>> {
  invariant(
    isNumber(expireAfter) && expireAfter > 0,
    "createExpireTransformConfig: `expireAfter` expected to be a positive number.",
  );

  invariant(
    isFunction(now),
    "createExpireTransformConfig: `now` expected to be a function.",
  );

  return {
    in(state) {
      return { t: now(), r: state };
    },

    out(plain) {
      return !plain || !plain.t || plain.t + expireAfter < now()
        ? null
        : plain.r;
    },
  };
}
