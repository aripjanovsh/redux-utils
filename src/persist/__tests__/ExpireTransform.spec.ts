import { createExpireTransformConfig } from "../ExpireTransform";

describe("ExpireTransform", () => {
  test("init validation", () => {
    expect(() => {
      // @ts-ignore
      createExpireTransformConfig(null);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      createExpireTransformConfig(undefined);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      createExpireTransformConfig(10, null);
    }).toThrowErrorMatchingSnapshot();
  });

  test("transformation", () => {
    let time = 1000;
    const now = () => time;
    const config = createExpireTransformConfig(1000, now);

    expect(config.in("foo")).toEqual({ r: "foo", t: 1000 });
    expect(config.out({ r: "foo", t: 1000 })).toBe("foo");

    time += 1000;

    expect(config.out({ r: "foo", t: 1000 })).toBe("foo");

    time += 500;

    expect(config.out({ r: "foo", t: 1000 })).toBe(null);
  });
});
