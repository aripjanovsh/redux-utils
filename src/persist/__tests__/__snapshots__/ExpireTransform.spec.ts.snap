// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpireTransform init validation 1`] = `"createExpireTransformConfig: \`expireAfter\` expected to be a positive number."`;

exports[`ExpireTransform init validation 2`] = `"createExpireTransformConfig: \`expireAfter\` expected to be a positive number."`;

exports[`ExpireTransform init validation 3`] = `"createExpireTransformConfig: \`now\` expected to be a function."`;
